src/ol_infrastructure/components/services/vault.py

```diff
@@ -1,5 +1,104 @@
-import pulumi
-from pulumi_vault import database
+"""
+This module controls mounting and configuration of database secret backends in Vault.
 
+This includes:
+- Mount a database backend at the configured mount point
+- Set the configuration according to the requirements of the backend type
+- Control the lease TTL settings
+- Define the base set of roles according to our established best practices
+"""
+from enum import Enum
+from typing import Text
 
-class VaultDatabaseRole(pulumi.ComponentResource):
+from pulumi import ComponentResource, ResourceOptions
+from pulumi_vault import Mount, database
+from pydantic import BaseModel, SecretStr
+
+from ol_infrastructure.lib.vault import (
+    mysql_sql_statements,
+    postgres_sql_statements
+)
+
+DEFAULT_PORT_POSTGRES = 5432
+DEFAULT_PORT_MYSQL = 3307
+DEFAULT_POT_MONGODB = 27017
+SIX_MONTHS = 60 * 60 * 24 * 30 * 6
+
+class DBEngines(str, Enum):
+    """Constraints for valid engine types that are supported by this component."""
+    postgres = 'postgresql'
+    mariadb = 'mysql'
+    mysql = 'mysql'
+    mysql_rds = 'mysql_rds'
+    mongodb = 'mongodb'
+
+class OLVaultDatabaseConfig(BaseModel):
+    db_name: Text
+    mount_point: Text
+    db_admin_username: Text
+    db_admin_password: SecretStr
+    verify_connection: bool = True
+    db_host: Text
+    max_ttl: int = SIX_MONTHS
+    default_ttl: int = SIX_MONTHS
+
+
+class OLVaultPostgresDatabaseConfig(OLVaultDatabaseConfig):
+    db_port: Int = DEFAULT_PORT_POSTGRES
+    db_connection: Text = 'postgresql://{{db_user}}:{{db_pass}}@{{db_host}}:{{db_port}}/{{db_name}}'
+    db_type: Text = DBEngines.postgres.value
+    role_statements: Dict[Text, Text] = postgresql_sql_statements
+
+
+class OLVaultMysqlDatabaseConfig(OLVaultDatabaseConfig):
+    db_port: Int = DEFAULT_PORT_MYSQL
+    db_connection: Text = '{{db_user}}:{{db_pass}}@tcp({{db_host}}:{{db_port}})/'
+    db_type: Text = DBEngines.mysql_rds.value
+    role_statements: Dict[Text, Text] = postgresql_sql_statements
+
+
+class OLVaultDatabaseBackend(ComponentResource):
+
+    def __init__(self, db_config: OLVaultDatabaseConfig, opts: ResourceOptions = None):
+        super().__init__('ol:services:VaultDatabaseBackend', db_config.db_name, None, opts)
```

Comment:
Not sure I follow where this is coming from

Reply:
That's part of the pattern for a component resource. It requires a call to super https://www.pulumi.com/docs/intro/concepts/programming-model/#authoring-a-new-component